src/ucp/tag/rndv.c

```diff
@@ -52,20 +62,85 @@ size_t ucp_tag_rndv_pack_rkey(ucp_request_t *sreq, ucp_lane_index_t lane,
 
     /* Check if the sender needs to register the send buffer -
      * is its datatype contiguous and does the receive side need it */
-    if (ucp_ep_rndv_md_flags(ep) & UCT_MD_FLAG_NEED_RKEY) {
+    if (ucp_ep_rndv_md_flags(ep, 0) & UCT_MD_FLAG_NEED_RKEY) {
         status = ucp_request_send_buffer_reg(sreq, lane);
         ucs_assert_always(status == UCS_OK);
 
         /* if the send buffer was registered, send the rkey */
         UCS_PROFILE_CALL(uct_md_mkey_pack, ucp_ep_md(ep, lane),
-                         sreq->send.state.dt.dt.contig.memh, rkey_buf);
+                         sreq->send.state.dt.dt.contig[0].memh, rkey_buf);
         *flags |= UCP_RNDV_RTS_FLAG_PACKED_RKEY;
         return ucp_ep_md_attr(ep, lane)->rkey_packed_size;
     }
 
     return 0;
 }
 
+static size_t ucp_tag_rndv_pack_rkeys(ucp_request_t *sreq, void *rkey_buf, uint16_t *flags)
+{
+    ucp_ep_t        *ep      = sreq->send.ep;
+    ucp_dt_state_t  *state   = &sreq->send.state.dt;
+    size_t           packet  = 0;
+    int              cnt     = 0;
+    int              i;
+    ucp_lane_index_t lane;
+
+    ucs_assert(UCP_DT_IS_CONTIG(sreq->send.datatype));
+
+    cnt = ucp_request_rndv_reg(sreq);
+    ucs_assert_always(cnt <= UCP_MAX_RNDV_LANES);
+
+    if (cnt) {
+        for (i = 0; i < ucp_ep_rndv_num_lanes(ep); i++) {
+            ucs_assert_always(!ucp_dt_is_empty_rndv_lane(state, i));
+
+            lane = ucp_ep_get_rndv_get_lane(ep, i);
+
+            if (ucp_ep_rndv_md_flags(ep, i) & UCT_MD_FLAG_NEED_RKEY) {
+                UCS_PROFILE_CALL(uct_md_mkey_pack, ucp_ep_md(ep, lane),
+                                 state->dt.contig[i].memh, rkey_buf + packet);
+                packet += ucp_ep_md_attr(ep, lane)->rkey_packed_size;
+            }
+        }
+        *flags |= UCP_RNDV_RTS_FLAG_PACKED_RKEY;
+    }
+
+    ucs_assert_always(packet <= ucp_ep_config(ep)->am.max_bcopy);
+
+    return packet;
+}
+
+static void ucp_tag_rndv_unpack_rkeys(ucp_request_t *req, void *rkey_buf, uint16_t flags)
+{
+    ucp_ep_t        *ep     = req->send.ep;
+    size_t           packet = 0;
+    ucp_lane_index_t rkeys;
+    ucp_lane_index_t lane;
+    int              i;
+
+    ucs_assert(UCP_DT_IS_CONTIG(req->send.datatype));
+
+    if (flags & (UCP_EP_FLAG_TAG_OFFLOAD_ENABLED | UCP_RNDV_RTS_FLAG_PACKED_RKEY)) {
```

Comment:
did you mean UCP_RNDV_RTS_FLAG_OFFLOAD here?
and why can't we use multi-rail in case of OFFLOAD? it's completed in SW anyway.. @brminich what do you think?

Reply:
tag offloading has own way to initialize data and it could not be used with multirail (just because there used different lanes). that is why tag offloading is processed separately

Reply:
do you mean because RTS has a key only for one lane?